Introduce automated comparison and PR reporting for G code changes

aepete opened the pull request:
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

This change introduces automated comparison & pr-reporting for G code changes.

- Note 1: this initial implementation only reports a textual description of the G code diff to PRs; a follow-up change will post diff images.
- Note 2: as this tooling comes online, I've configured it to be opt-in (off by default).  It is enabled by including the keyword `RUN_G_DIFF_CHECK` anywhere in the PR description -- hereby having enabled it for this PR ;)

### Why should this Pull Request be merged?

Without such tooling, the only options to understand G code (binary) changes are to either open / diff the LV files by-hand and/or have screenshots be attached to the PR (also a manual process).

### What testing has been done?

Ensured proper operation of workflow in a predecessor draft PR (#390)
Ensured proper operation of python & LabVIEW code
- Numerous executions of local dev workflow using `run_diff.bat`
- Numerous executions of PR workflow via earlier draft PR as well as this PR.

aepete commented on Source/Tools/run_diff.vi:
Hopefully this will be one of the last times I have to manually post screen captures ;)
![image](https://github.com/ni/measurementlink-labview/assets/88401200/431c6c65-4e07-45ff-82fa-44b1524bd9c7)
![image](https://github.com/ni/measurementlink-labview/assets/88401200/690b7cf7-1e71-4774-b4d5-cf3af1d911d8)

Notes
- The overall flow of this file roughly resembles run_tests.vi
- Basic sequence, left-to-right:
   1. Parse args & prepare to diff
   2. Process each affected file as appropriate (currently only outputting text content for each)
   3. Package, format, and return results

aepete commented on Source/Tools/subVIs/Parse Diff Report to Text.vi:
This utility VI is only slightly modified from its [predecessor, from NI Veristand](https://github.com/ni/niveristand-custom-device-build-tools/blob/main/lv/operations/DiffVI/Modules/Screenshots/ParseDiffXMLtoPlainText.vi).

dixonjoel commented:
@aepete Could you add some sample text for a typical modified VI for us to get a sense of what the diffs look like? Would all of that info go in the comment here e.g. where we see
> 3 of 3 files processed successfully.
Processed run_diff.vi: Success
Added run_diff.vi.  Screenshots coming soon.

> Processed Change Type.ctl: Success
Added Change Type.ctl.  Screenshots coming soon.

> Processed Parse Diff Report to Text.vi: Success
Added Parse Diff Report to Text.vi.  Screenshots coming soon.

aepete commented:
> @aepete Could you add some sample text for a typical modified VI

Yes -- below I've listed two examples of how this might look, first for a simple VI modification and second for a larger modification:

------------------------
Processed `Get Default for Type.vi`: Success
Attribute, Panel, and Diagram changes (screenshots coming soon):
Attribute Changes
&nbsp;&nbsp;&nbsp;&nbsp; None

Front Panel Functional Changes
&nbsp;&nbsp;&nbsp;&nbsp; None

Block Diagram Functional Changes
&nbsp;&nbsp;&nbsp;&nbsp; Added
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;String Constant password ("") Added at (-12,194)
&nbsp;&nbsp;&nbsp;&nbsp; wiring changes

------------------------

Processed `Run Service.vi`: Success
Attribute, Panel, and Diagram changes (screenshots coming soon):
Attribute Changes
&nbsp;&nbsp;&nbsp;&nbsp; Clear Indicators When Called

Front Panel Functional Changes
&nbsp;&nbsp;&nbsp;&nbsp; Added
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Boolean Debug? Added at (24,279)
&nbsp;&nbsp;&nbsp;&nbsp; control/indicator assignment
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Connector Pane&nbsp;&nbsp;Changed.

Block Diagram Functional Changes
&nbsp;&nbsp;&nbsp;&nbsp; Added
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Case Selector&nbsp;&nbsp;Added at (-1083,-8)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Case Selector&nbsp;&nbsp;Added at (-806,-81)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Case Structure&nbsp;&nbsp;Added at (-1083,-58)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Case Structure&nbsp;&nbsp;Added at (-806,-159)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Front Panel Terminal Debug? Added at (-1532,-125)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;String Constant address Added at (-1057,-20)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SubVI Clear Errors.vi Added at (-288,-15)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SubVI Clear Errors.vi Added at (50,39)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SubVI MeasurementLink Measurement Server.lvlib:Get Discovery Service Address.vi Added at (-1061,-15)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1006,-15)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1006,-15)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1006,-15)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1006,9)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1006,9)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1006,9)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1083,9)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1083,9)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-1083,9)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-751,-127)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-751,-127)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-761,-62)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-761,-62)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-806,-127)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-806,-127)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-984,-79)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tunnel&nbsp;&nbsp;Added at (-984,-79)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Type Definition gRPC ID Added at (-781,-89)
&nbsp;&nbsp;&nbsp;&nbsp; Deleted
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SubVI MeasurementLink Discovery.lvlib:Get Discovery Service Address.vi Deleted at (-1070,-15)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SubVI MeasurementLink Measurement Server.lvlib:Validate Metadata.vi Deleted at (-1133,-15)
&nbsp;&nbsp;&nbsp;&nbsp; wiring changes

aepete commented:
> Yes -- below I've listed two examples

I tend to find the "Tunnel Added" and "Case Selector Added" line items to be relatively unhelpful.  I could filter them out here, but I think I'd prefer to focus on making efficiently-rendered images of the graphical diff as our next step.

bkeryan commented on .github/workflows/run_pr_commenter.yml:
```diff
@@ -0,0 +1,27 @@
+    - name: Run G Diff
+      id: run-diff
+      run: |
+        python ${{runner.workspace}}\measurementlink-labview\Source\Tools\run_diff.py --token ${{secrets.GITHUB_TOKEN}} --pull-req ${{github.event.pull_request.number}}
```
Use forward slashes in GH workflows for portability, unless the app absolutely requires backslashes (Python does not)

aepete replied:
Updated to forward slashes in latest update.  Also corrected to use forward slashes in  `run_g_tests`

bkeryan commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,13 @@
+  cancel-in-progress: true
+
+jobs:
+  run_pr_commenter:
```
Consider adding friendly names to these steps because their names show up in the "checks" box/tab.

aepete replied:
User-friendly names included in Latest Update:

![image](https://github.com/ni/measurementlink-labview/assets/88401200/53dcd1d9-d501-4a81-b3af-8ac991a0c61f)

bkeryan commented on .github/workflows/pr.yml:
(BTW, in measurementlink-python, pr.yml calls ci.yml or vice-versa, but as long as ci.yml and pr.yml use workflow calls so that the real workflow logic is single-sourced, I don't think duplication between ci.yml and pr.yml is a big deal.)

aepete replied:
Changes applied in latest update.

pbirkhol-ni commented on Source/Tools/run_diff.vi:
This VI seems to support a --renamed_labview_files command line arg but I don't think we ever pass that arg.

pbirkhol-ni replied:
I now see that the LabVIEW code says "coming soon".

bkeryan commented on .gitignore:
```diff
@@ -14,6 +14,10 @@
\ No newline at end of file
+/Build Output/
+
+# Python Cache
+/Source/Tools/__pycache__/
```
measurementlink-python has a more general version of this:

```suggestion
# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
*$py.class
```

pbirkhol-ni commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,21 @@
+
+  run_pr_commenter:
+    name: Run PR Commenter
+    if: ${{contains(github.event.pull_request.body, 'RUN_G_DIFF_CHECK') && (github.repository == 'ni/measurementlink-labview' || github.event.pull_request.head.repo.full_name == 'ni/measurementlink-labview') }}
```
I think having this opt-in is a good idea and we might consider leaving it opt-in permanently. It isn't that uncommon for us to make a change that touches hundreds of VIs (hopefully that will be less common going forward). In that case, I don't think I'd want this tool to run. Maybe we can add a comment to the PR template like this:

`<!-- RUN_G_DIFF_CHECK: Uncomment this line to enable VI diffing. Consider leaving VI diffing disable for PRs that modify a lot of VIs  -->`

And then you can modify the logic in `pr.yml` to decide whether to run the differ. This seems like the best of all worlds.

bkeryan replied:
It might be tricky to check whether text is commented out. I don't know if `github.event.pull_request.body` strips out commented out text.

Consider using a checkbox like Renovate does: https://github.com/ni/measurementlink-labview/pull/407

markdown:
` - [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check this box`

rendered:
 - [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check this box

bkeryan replied:
When you check/uncheck a checkbox, GitHub automatically edits the PR body, reply, or whatever.

pbirkhol-ni replied:
Yeah, the checkbox is a good idea.

aepete replied:
Tracking this with a follow-up task: 2582143

pbirkhol-ni reviewed (approved):
I didn't review the nitty gritty details. The overall idea and implementation look good to me!

aepete commented on Source/Tools/run_diff.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def parse_options(args):
+    import optparse
```
ahh, just copied it over from an old script.  Should be easy to replace with `argparse`

bkeryan replied:
FYI: https://dev.azure.com/ni/DevCentral/_wiki/wikis/AppCentral.wiki/35648/Command-Line-Interface-Parser

aepete replied:
Cool, I'll go with `click`

aepete replied:
Fixed in latest update.

bkeryan reviewed (approved):
click changes are ok
